### PR TITLE
Memory leaks

### PR DESCRIPTION
**Pull request by catap**
I found and fix few memory leaks, meanwhile my tests shown that it isn't all :(

### CODE REVIEW

File: src/lib/reader.c [hunk 1]
```diff
@@ -203,6 +203,7 @@ pgp_reader_push(pgp_stream_t *          stream,
 
     if ((readinfo = calloc(1, sizeof(*readinfo))) == NULL) {
         (void) fprintf(stderr, "pgp_reader_push: bad alloc\n");
+        return 0;
```
Comment (ronaldtse) [inline src/lib/reader.c]: I think where places we just `return` in an if-else block should just be short circuited like this?

```c
if (x) {
  ...
  return 0;
} else {
  ...
}
```

To:
```c
if (x) {
  ...
  return 0;
}
...
```

File: src/lib/signature.c [hunk 2]
```diff
@@ -1257,11 +1257,18 @@ pgp_sign_detached(rnp_ctx_t *    ctx,
 
     /* create a new signature */
     sig = pgp_create_sig_new();
+    if (sig == NULL) {
```
Comment (ronaldtse) [inline src/lib/signature.c]: Can this be handled inside `pgp_create_sig_new`?

File: src/lib/writer.c [hunk 3]
```diff
@@ -252,12 +252,14 @@ pgp_writer_set(pgp_output_t *          output,
 {
     if (output->writer.writer) {
         (void) fprintf(stderr, "pgp_writer_set: already set\n");
+        return 0;
     } else {
```
Comment (ronaldtse) [inline src/lib/writer.c]: Can short-circuit?

File: src/lib/writer.c [hunk 4]
```diff
@@ -272,19 +274,24 @@ pgp_writer_set(pgp_output_t *          output,
...
                 pgp_writer_finaliser_t *finaliser,
                 pgp_writer_destroyer_t *destroyer,
                 void *                  arg)
 {
-    pgp_writer_t *copy;
+    pgp_writer_t *copy = NULL;
 
     if ((copy = calloc(1, sizeof(*copy))) == NULL) {
         (void) fprintf(stderr, "pgp_writer_push: bad alloc\n");
+        return 0;
```
Comment (ronaldtse) [inline src/lib/writer.c]: Can short-circuit?

File: src/lib/writer.c [hunk 5]
```diff
@@ -272,19 +274,24 @@ pgp_writer_set(pgp_output_t *          output,
...
     } else if (output->writer.writer == NULL) {
+        if (copy != NULL) {
+            free(copy);
+        }
         (void) fprintf(stderr, "pgp_writer_push: no orig writer\n");
+        return 0;
```
Comment (ronaldtse) [inline src/lib/writer.c]: Can short-circuit?

File: src/lib/packet-parse.c [hunk 6]
```diff
@@ -1421,8 +1431,10 @@ parse_userattr(pgp_region_t *region, pgp_stream_t *stream)
 void
 pgp_userid_free(uint8_t **id)
 {
-    free(*id);
-    *id = NULL;
+    if (*id != NULL) {
```
Comment (ronaldtse) [inline src/lib/packet-parse.c]: Can short-circuit to this?

```c
if (*id == NULL)
  return;

free(*id);
*id = NULL;
```

File: src/lib/signature.c (hunk 2 above)
Comment (catap) [inline src/lib/signature.c]: no, because pgp_create_sig_new is a just calloc wrapper

Comment (ronaldtse) [inline src/lib/signature.c]: Thank you @catap !

**Comment by ronaldtse** [comment]
Thanks @catap ! I think we need one more reviewer... @dewyatt / @ni4 ?

File: src/lib/packet-parse.c [hunk 7]
```diff
@@ -888,8 +889,10 @@ string_free(char **str)
 void
 pgp_subpacket_free(pgp_subpacket_t *packet)
 {
-    free(packet->raw);
-    packet->raw = NULL;
+    if (packet->raw) {
```
Comment (ronaldtse) [inline src/lib/packet-parse.c]: @catap can we short circuit this too? Thanks!

File: src/lib/misc.c [hunk 8]
```diff
@@ -964,7 +964,9 @@ pgp_reader_push_sum16(pgp_stream_t *stream)
     if ((arg = calloc(1, sizeof(*arg))) == NULL) {
```
Comment (ronaldtse) [inline src/lib/misc.c]: @catap can we short circuit this whole function? Thanks.

Comment (catap) [inline src/lib/misc.c]: Fixed at https://github.com/riboseinc/rnp/pull/259/commits/2f0533965cddca7ad48ec7b354fd1b62d0df3840

File: src/lib/packet-parse.c (hunk 7 above)
Comment (catap) [inline src/lib/packet-parse.c]: Fixed at https://github.com/riboseinc/rnp/pull/259/commits/2f0533965cddca7ad48ec7b354fd1b62d0df3840

**Comment by ronaldtse** [comment]
Looks good and going to merge this now. Thanks @catap !

**Pull request merged**